### PR TITLE
Add comment target flag and parsing

### PR DESCRIPTION
**Pull request by tasdomas**
One significant change in this PR is that bitbucket commit comments will no longer be automatically posted at the PR level.

The `--target=auto` scenario is not yet being handled in this PR, this will be implemented in a follow-up.

The general idea is as follows:

- legacy comment target flags (`--pr`, `--commitSha`) take precendence.
- the `--issue` flag will be removed in a follow-up (replaced with `--target=issue#id`)
- the `--target` flag can take both qualified (`--target=pr#id` or `--target=commit#sha` or `--target=issue#id`) and unqualified values (`--target=pr` or `--target=commit`)
- when parsing an unqualified `pr` target, cml will first attempt to determine the PR number using the forge context (env vars for bitbucket and gitlab, `context.pull_request.number` for github) - if that fails, an attempt will be made to determine the pr number using the sha of the head commit
- when parsing an unqualified `commit` target, comments will be attached to the head commit

### CODE REVIEW

**Comment by dacbd** [comment]
This portion is to just add the `--target=commit` type functionality? and not `--target=issue#1234` type correct?

**Comment by tasdomas** [comment]
This adds support for all target types

**Review by dacbd — CHANGES_REQUESTED** [review]
using a `--target=issue#1234` creates commit comment. see: https://github.com/iterative/cml-playground/actions/runs/3388499291/jobs/5630555691#step:5:11

**Comment by tasdomas** [comment]
> using a `--target=issue#1234` creates commit comment. see: https://github.com/iterative/cml-playground/actions/runs/3388499291/jobs/5630555691#step:5:11

Fixed.

**Review by DavidGOrtega — CHANGES_REQUESTED** [review]
> One significant change in this PR is that bitbucket commit comments will no longer be automatically posted at the PR level.

Why? It goes against the current CML definition (the video is even showing it as the rest of forges)

**Comment by tasdomas** [comment]
> > One significant change in this PR is that bitbucket commit comments will no longer be automatically posted at the PR level.
> 
> Why? It goes against the current CML definition (the video is even showing it as the rest of forges)
Because this behavior is inconsistent across forges. We don't do the same thing for github, where the support for commit comments is similar to bitbucket's. In addition to this, the side effect of the current implementation is that the PR comments are augmented with a link to the commit on all forges, which imho is unnecessary.

File: bin/cml/comment/create.test.js [hunk 1]
```diff
@@ -17,11 +17,16 @@ describe('Comment integration tests', () => {
         --help    Show help                                                  [boolean]
 
       Options:
+        --target                    Forge object to create comment on, can be one of
+                                    pr, commit or issue. Specify 'issue#123' to create
```
Comment (casperdcl) [inline bin/cml/comment/create.test.js:21]: what syntaxes are supported?

- `123`
- `#123`
- `issue#123`
- `pr#123`
- `mr#123`

File: bin/cml/comment/create.js [hunk 2]
```diff
@@ -18,21 +18,23 @@ exports.builder = (yargs) =>
     .options(exports.options);
 
 exports.options = kebabcaseKeys({
+  target: {
+    type: 'string',
+    description:
+      'Forge object to create comment on, can be one of pr, commit or issue. ' +
+      "Specify 'issue#123' to create a comment on a specific issue. " +
```
Comment (casperdcl) [inline bin/cml/comment/create.js:25]: what syntaxes are supported?

- `123`
- `#123`
- `issue#123`
- `pr#123`
- `mr#123`

Comment (tasdomas) [reply bin/cml/comment/create.js:25]: Valid targets are:
- `auto` determine comment target automatically:
  - PR if running in a PR context
  - PR if HEAD commit is part of a PR
  - HEAD commit (fallback)
- `pr` - determine pr by context or head commit (making allowances for how github's PR merge commits)
- `pr#123` - specific PR number
- `commit` - current HEAD commit
- `commit#adfa12` - specific commit
- `issue#123` - specific issue

File: bin/cml/comment/create.js [hunk 3]
```diff
@@ -18,21 +18,23 @@ exports.builder = (yargs) =>
...
+      'By default cml will create a PR comment if running in a forge PR-related action ' +
+      'or if HEAD is in a PR branch. Otherwise a commit comment will be created.'
+  },
   pr: {
     type: 'boolean',
     description:
       'Post to an existing PR/MR associated with the specified commit',
     conflicts: ['issue']
```
[the remark this answers is not in the document]
Comment (tasdomas) [reply bin/cml/comment/create.js]: Fixed, thank you

**Comment by tasdomas** [comment]
> > One significant change in this PR is that bitbucket commit comments will no longer be automatically posted at the PR level.
> 
> Why? It goes against the current CML definition (the video is even showing it as the rest of forges)

Can we close this change request?

File: src/drivers/bitbucket_cloud.js [hunk 4]
```diff
@@ -536,6 +546,10 @@ class BitbucketCloud {
 
     return responseBody;
   }
+
+  warn(message) {
+    console.error(message);
+  }
```
Comment (DavidGOrtega) [inline src/drivers/bitbucket_cloud.js:552]:
```suggestion
```

File: src/drivers/gitlab.js [hunk 5]
```diff
@@ -536,6 +546,10 @@ class Gitlab {
 
     return await response.json();
   }
+
+  warn(message) {
+    console.error(message);
+  }
 }
```
Comment (DavidGOrtega) [inline src/drivers/gitlab.js:553]:
```suggestion
```

**Review by DavidGOrtega — CHANGES_REQUESTED** [review]
👋 great effort!
I do not get the `warn` method.
It should be at most a CML capability and to be honest not even that. We already have a logging mechanism in place.
Its also missing in GH?

Probably I missing something 🙏

File: src/drivers/bitbucket_cloud.js (hunk 4 above)
Comment (0x2b3bfa0) [reply src/drivers/bitbucket_cloud.js:552]: Perhaps use `winston.warn` instead?

```suggestion

  warn(message) {
    winston.warn(message);
  }
```

File: src/drivers/gitlab.js (hunk 5 above)
Comment (0x2b3bfa0) [reply src/drivers/gitlab.js:553]: Perhaps use `winston.warn` instead?

```suggestion

  warn(message) {
    winston.warn(message);
  }
}
```

**Comment by DavidGOrtega** [comment]
> The warn method is meant to display warnings in the forge user interface, rather than just in the logs. 

But its doing at the end a console.log so probably winston can handle it?

File: src/cml.js [hunk 6]
```diff
@@ -162,16 +163,15 @@ class CML {
   }
 
   async commentCreate(opts = {}) {
-    const triggerSha = await this.triggerSha();
```
[the remark this answers is not in the document]
Comment (tasdomas) [reply src/cml.js:165]: It's considered in `parseCommentTarget`

File: src/cml.js [hunk 7]
```diff
@@ -162,16 +163,15 @@ class CML {
...
     const {
-      commitSha: inCommitSha = triggerSha,
-      issue: issueId,
+      commitSha: inCommitSha,
```
[the remark this answers is not in the document]
Comment (tasdomas) [reply src/cml.js:167]: Good point, thank you

File: src/commenttarget.js [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+  let commitPr;
+  switch (commentTarget) {
+    case 'commit':
+      return { target: 'commit', commitSha: drv.sha };
+    case 'pr':
+    case 'auto':
+      // Determine PR id from forge env vars (if we're in a PR context).
+      prId = drv.pr;
+      if (prId) {
+        return { target: 'pr', prNumber: prId };
```
[the remark this answers is not in the document]
Comment (tasdomas) [reply src/commenttarget.js:29]: Fixed, thank you

File: src/cml.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (tasdomas) [reply src/cml.js]: Fixed in https://github.com/iterative/cml/pull/1241/commits/e7a7996b731956e891a45087a0d1ccad611ff6d7

File: src/commenttarget.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (tasdomas) [reply src/commenttarget.js]: Fixed in https://github.com/iterative/cml/pull/1241/commits/e7a7996b731956e891a45087a0d1ccad611ff6d7

File: src/drivers/bitbucket_cloud.js (hunk 4 above)
Comment (tasdomas) [reply src/drivers/bitbucket_cloud.js]: Fixed in https://github.com/iterative/cml/pull/1241/commits/e7a7996b731956e891a45087a0d1ccad611ff6d7

**Review by DavidGOrtega — APPROVED** [review]

**Review by 0x2b3bfa0 — APPROVED** [review]

**Review by dacbd — APPROVED** [review]
final testing to be done in the feature branch

File: bin/cml/comment/create.js (hunk 2 above)
Comment (casperdcl) [reply bin/cml/comment/create.js:25]: I think we need:

- `#123` to auto-infer pr-or-issue
- `mr` alias for `pr`
- `mr#123` alias for `pr#123`
- why do we have `commit#d34db3`? Doesn't `--commit-sha=d34db3` already cover this?
- full URL (e.g. if they want to target another PR returned by `cml pr`)

**Pull request merged**